src/sdk/ConversationTranslator/Conversation.ts

```diff
@@ -38,124 +52,612 @@ export abstract class Conversation implements IConversation {
+
+        } catch (error) {
+            if (!!err) {
+                if (error instanceof Error) {
+                    const typedError: Error = error as Error;
+                    err(typedError.name + ": " + typedError.message);
+
+                } else {
+                    err(error);
+                }
+            }
+        }
+    }
+
+    /**
+     * Starts a new conversation as host.
+     * @param cb
+     * @param err
+     */
     public startConversationAsync(cb?: () => void, err?: (e: string) => void): void {
 
-        throw new Error("Method not implemented.");
+        try {
+
+            // connect
+            this.privConnection = new ConversationConnection(this.privConfig);
+            this.privConnection.connect(this.privRoom);
+
+            if (!!cb) {
+                try {
+                    cb();
+                } catch (e) {
+                    if (!!err) {
+                        err(e);
+                    }
+                }
+                cb = undefined;
+            }
+
+        } catch (error) {
+            if (!!err) {
+                if (error instanceof Error) {
+                    const typedError: Error = error as Error;
+                    err(typedError.name + ": " + typedError.message);
+
+                } else {
+                    err(error);
+                }
+            }
+        }
+    }
+
+    /**
+     * Join a conversation as a participant.
+     * @param conversation
+     * @param nickname
+     * @param lang
+     * @param cb
+     * @param err
+     */
+    public joinConversationAsync(conversationId: string, nickname: string, lang: string, cb?: (result: any) => void, err?: (e: string) => void): void {
+
+        try {
+
+            this.reset();
+
+            // join the conversation
+            this.privManager.createOrJoin(this.privProperties, conversationId,
+                ((room: IInternalConversation) => {
+
+                    Contracts.throwIfNullOrUndefined(room, "invalid conversation");
+                    this.privRoom = room;
+                    this.privConfig.authorizationToken = room.cognitiveSpeechAuthToken;
+
+                    // join callback
+                    if (!!cb) {
+                        cb(room);
+                    }
+
+                }),
+                ((error: any) => {
+                    if (!!err) {
+                        if (error instanceof Error) {
+                            const typedError: Error = error as Error;
+                            err(typedError.name + ": " + typedError.message);
+
+                        } else {
+                            err(error);
+                        }
+                    }
+                }));
+
+        } catch (error) {
+            if (!!err) {
+                if (error instanceof Error) {
+                    const typedError: Error = error as Error;
+                    err(typedError.name + ": " + typedError.message);
+
+                } else {
```

Comment:
Nitpick: have you considered moving this error handling code into a common method? It looks like this logic is repeated in many places

Reply:
Added a shared function for the callbacks.